cmd/ztest/ztest.c

```diff
@@ -7052,6 +7119,379 @@ ztest_import(ztest_shared_t *zs)
+		 */
+		root = make_vdev_root(newpath, NULL, NULL, csize, ashift, NULL,
+		    0, 0, 1);
+		/*
+		 * Now attach the vdev to the raidz so it will expand
+		 */
+		if (ztest_opts.zo_verbose >= 1) {
+			(void) printf("expanding raidz\n");
+		}
+		error = spa_vdev_attach(spa, rzvd->vdev_guid, root, B_FALSE,
+		    B_FALSE);
+		nvlist_free(root);
+		if (error != 0) {
+			fatal(0, "raidz expand: attach (%s %llu) returned %d",
+			    newpath, csize, error);
+		}
+
+	} else { /* check the pool is healthy */
+		/*
+		 * Set pool check done flag, main program will run azdb check of
+		 * the pool when we exit.
+		 */
+		ztest_shared_opts->zo_raidz_expand_test = (UINT64_MAX - 1);
+		/* XXX - wait for reflow done? */
+		if (ztest_opts.zo_verbose >= 1) {
+			(void) printf("\nverifying raidz expansion\n");
+			if (ztest_opts.zo_verbose > 1) {
+				/* XXX - pause to allow debugger attach */
+				(void) printf(
+				    "our pid is %d, pausing for 10 seconds\n",
+				    getpid());
+				sleep(10);
+			}
+		}
+		VERIFY0(ztest_scrub_impl(spa));
+		if (ztest_opts.zo_verbose >= 1) {
+			(void) printf("raidz expansion scrub check complete\n");
+		}
+		goto done;
+	}
+
+	/*
+	 * Wait for desired reflow offset to be reached and kill the test
+	 */
+	/*
+	 * Wait for reflow to begin
+	 */
+	while (spa->spa_raidz_expand == NULL) {
+		txg_wait_synced(spa_get_dsl(spa), 0);
+		(void) poll(NULL, 0, 100); /* wait 1/10 second */
+	}
+	spa_config_enter(spa, SCL_CONFIG, FTAG, RW_READER);
+	(void) spa_raidz_expand_get_stats(spa, pres);
+	spa_config_exit(spa, SCL_CONFIG, FTAG);
+	while (pres->pres_state != DSS_SCANNING) {
+		txg_wait_synced(spa_get_dsl(spa), 0);
+		(void) poll(NULL, 0, 100); /* wait 1/10 second */
+		spa_config_enter(spa, SCL_CONFIG, FTAG, RW_READER);
+		(void) spa_raidz_expand_get_stats(spa, pres);
+		spa_config_exit(spa, SCL_CONFIG, FTAG);
+	}
+
+	ASSERT3U(pres->pres_state, ==, DSS_SCANNING);
+	ASSERT3U(pres->pres_to_reflow, !=, 0);
+	/*
+	 * Set so when we are killed we go to raidz checking rather than
+	 * restarting test.
+	 */
+	ztest_shared_opts->zo_raidz_expand_test = UINT64_MAX; /* pass 1 done */
+	if (ztest_opts.zo_verbose >= 1) {
+		(void) printf("raidz expandsion reflow started,"
+		    " waiting for offset %llu to be reched\n",
+		    (u_longlong_t)desreflow);
+	}
+
+	while (pres->pres_reflowed < desreflow) {
+		txg_wait_synced(spa_get_dsl(spa), 0);
+		(void) poll(NULL, 0, 100); /* wait 1/10 second */
+		spa_config_enter(spa, SCL_CONFIG, FTAG, RW_READER);
+		(void) spa_raidz_expand_get_stats(spa, pres);
+		spa_config_exit(spa, SCL_CONFIG, FTAG);
+	}
+
+	/*
+	 * XXX - should we clear the reflow pause here?
+	 */
+	if (ztest_opts.zo_verbose >= 1) {
+		(void) printf("killing raidz expandsion test offset at %llu\n",
+		    (u_longlong_t)pres->pres_reflowed);
+	}
+	ztest_kill(zs);
+
+done:
+#if 0
+	/*
+	 * XXX - we don't have any threads here because we killed the test and
+	 * then ran a scrub in this thread rather than starting sub threads.
```

Comment:
it doesn't seem like the "goto" is necessary, we could move this code to the one place that we "goto done".

Reply:
Yeah, this is an artifact of copying ztest_run and hacking it around.  will clean up.